### PR TITLE
image captioning using CNN and Transformers

### PR DESCRIPTION
**Pull request by AakashKumarNain**

### CODE REVIEW

File: examples/vision/image_captioning.py [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+EMBED_DIM = 512
+# Number of self-attention heads
+NUM_HEADS = 2
+# Dimensions of the feed-forward network
+FF_DIM = 512
+
+# Other training parameters
+BATCH_SIZE = 32
+EPOCHS = 15
+AUTOTUNE = tf.data.experimental.AUTOTUNE
```
Comment (fchollet) [inline examples/vision/image_captioning.py:64]: `tf.data.AUTOTUNE`

File: examples/vision/image_captioning.py [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+# Path to the images
+IMAGES_PATH = "Flicker8k_Dataset"
+
+# Desired image dimensions
+IMAGE_SIZE = (299, 299)
+
+# Vocabulary size
+VOCAB_SIZE = 10000
+# Maximum length allowed for any sequence
+MAX_SEQ_LENGTH = 40
```
Comment (fchollet) [inline examples/vision/image_captioning.py:52]: It's not a max in this case, it's a fixed length. There can be no shorter sequence (they get padded). `SEQ_LENGTH`.

File: examples/vision/image_captioning.py [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+
+
+strip_chars = "!\"#$%&'()*+,-./:;<=>?@[\]^_`{|}~"
+strip_chars = strip_chars.replace("<", "")
+strip_chars = strip_chars.replace(">", "")
+
+vectorization = TextVectorization(
+    max_tokens=VOCAB_SIZE,
+    output_mode="int",
+    output_sequence_length=MAX_SEQ_LENGTH,
```
Comment (fchollet) [inline examples/vision/image_captioning.py:174]: This is the part that will cause the sequences to have a fixed length, btw.

File: examples/vision/image_captioning.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+def get_cnn_model():
+    base_model = efficientnet.EfficientNetB0(
+        input_shape=(*IMAGE_SIZE, 3),
+        include_top=False,
+        weights="imagenet",
+    )
+    # We freeze our feature extractor
+    base_model.trainable = False
+    base_model_out = base_model.output
+    base_model_out = layers.Reshape((-1, 1280))(base_model_out)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:239]: `Flatten`, no?

File: examples/vision/image_captioning.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+    cnn_model = keras.models.Model(base_model.input, base_model_out)
+    return cnn_model
+
+
+class TransformerEncoderBlock(layers.Layer):
+    def __init__(self, embed_dim, dense_dim, num_heads, **kwargs):
+        super(TransformerEncoderBlock, self).__init__(**kwargs)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:246]: Minor comment: you can switch to Python 3 style `super` calls since the code is intended to be Py3 only

File: examples/vision/image_captioning.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+        cnn_model,
+        encoder,
+        decoder,
+    ):
+        super(ImageCaptioningModel, self).__init__()
+        self.cnn_model = cnn_model
+        self.encoder = encoder
+        self.decoder = decoder
+        self.train_loss_tracker = keras.metrics.Mean(name="loss")
+        self.val_loss_tracker = keras.metrics.Mean(name="val_loss")
```
Comment (fchollet) [inline examples/vision/image_captioning.py:361]: You can use a single tracker for both

File: examples/vision/image_captioning.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+
+        # 7. Get the gradients
+        grads = tape.gradient(loss, train_vars)
+
+        # 8. Update the trainable weights
+        self.optimizer.apply_gradients(zip(grads, train_vars))
+
+        # 9. Update the loss tracker and return the loss value
+        self.train_loss_tracker.update_state(loss)
+        return {"loss": self.train_loss_tracker.result()}
```
Comment (fchollet) [inline examples/vision/image_captioning.py:398]: Why not keep track of another metric like accuracy?

File: examples/vision/image_captioning.py [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+caption_model = ImageCaptioningModel(
+    cnn_model=cnn_model, encoder=encoder, decoder=decoder
+)
+
+"""
+## Model training
+"""
+
+# Define the loss function
+ce = keras.losses.SparseCategoricalCrossentropy(from_logits=False)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:428]: Use fully-spelled out variable names

File: examples/vision/image_captioning.py [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+
+# EarlyStopping criteria
+es = keras.callbacks.EarlyStopping(patience=3, restore_best_weights=True)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:431]: Here too

File: examples/vision/image_captioning.py [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+        sampled_token = index_lookup[sampled_token_index]
+        if sampled_token == " <end>":
+            break
+        decoded_caption += " " + sampled_token
+
+    print("PREDICTED CAPTION:", end=" ")
+    print(decoded_caption.replace("<start> ", "").replace(" <end>", "").strip())
+
+
+# Check predictions for a few samples
```
Comment (fchollet) [inline examples/vision/image_captioning.py:483]: You may want to include a text block that features hardcoded example outputs, along with a comment saying that this is what we get after X epochs. It's often the case that we have to reduce the number of epochs in the example to keep the code easily runnable.

File: examples/vision/image_captioning.py [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+    Returns:
+        Traning and validation datasets as two separated dicts
+    """
+
+    # 1. Get the list of all image names
+    all_images = list(caption_data.keys())
+
+    # 2. Shuffle if necessary
+    if shuffle:
+        np.random.shuffle(all_images)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:127]: I recommend seeding. You might get wildly different results from run to run if you train on different splits every time.

**Review by fchollet — COMMENTED** [review]
Awesome stuff! 👍

File: examples/vision/image_captioning.py (hunk 2 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:52]: Ahh yes, I will fix this

File: examples/vision/image_captioning.py (hunk 5 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:246]: Totally

File: examples/vision/image_captioning.py (hunk 7 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:398]: Will update this

File: examples/vision/image_captioning.py (hunk 10 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:483]: Okay. I will add it

File: examples/vision/image_captioning.py (hunk 11 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:127]: I was using seed in my original experiment, it might have been removed accidentally while converting the notebook to py file and fromatting

**Comment by AakashKumarNain** [comment]
Thanks @fchollet for the review. I will update this

File: examples/vision/image_captioning.py (hunk 4 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:239]: Flatten will give a shape `(None, n)` but what we want is the shape `(None, m, n)` else the attention module will complain

**Comment by AakashKumarNain** [comment]
@fchollet Idk if it's the `black` version that is different in the CI but I format the scripts with `black` every time on my side before pushing the changes

File: examples/vision/image_captioning.py [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+"""
+## Download the dataset
+
+We will be using the Flickr8K dataset for this tutorial. This dataset comprises of more
+than 8,000 images that are each paired with five different captions.
+"""
+
+
+"""shell
+! wget -q https://github.com/jbrownlee/Datasets/releases/download/Flickr8k/Flickr8k_Dataset.zip
```
Comment (fchollet) [inline examples/vision/image_captioning.py:39]: Shell commands shouldn't have the !

File: examples/vision/image_captioning.py (hunk 11 above)
Comment (fchollet) [reply examples/vision/image_captioning.py:131]: Please add it back.

File: examples/vision/image_captioning.py [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+        # 1. Get image embeddings
+        img_embed = self.cnn_model(batch_img)
+
+        # 2. Pass image embeddings to encoder
+        encoder_out = self.encoder(img_embed, training=training)
+
+        # 3. Pass each of the five captions one by one the decoder
+        # along with the encoder outputs and compute the loss for each caption.
+        # We will also compute the accuracy
+        for i in range(5):
```
Comment (fchollet) [inline examples/vision/image_captioning.py:381]: 5 here should be a model constructor arg, most likely

File: examples/vision/image_captioning.py [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+            batch_seq_inp = batch_seq[:, i, :-1]
+            batch_seq_true = batch_seq[:, i, 1:]
+            batch_seq_pred = self.decoder(batch_seq_inp, encoder_out, training=training)
+            loss += self.loss(batch_seq_true, batch_seq_pred)
+            acc += keras.metrics.sparse_categorical_accuracy(
+                batch_seq_true, batch_seq_pred
+            )
+
+        return loss, acc / 5.0
```
Comment (fchollet) [inline examples/vision/image_captioning.py:390]: Likewise

File: examples/vision/image_captioning.py (hunk 11 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:131]: Added the seed for both `numpy` as well as `tf`

File: examples/vision/image_captioning.py [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+
+        self.pe = PositionalEmbedding(
+            embed_dim=EMBED_DIM, sequence_length=SEQ_LENGTH, vocab_size=VOCAB_SIZE
+        )
+        self.out = layers.Dense(VOCAB_SIZE, activation="softmax")
+        self.dropout = layers.Dropout(0.5)
+
+    def call(self, inputs, encoder_outputs, training, mask=None):
+        inputs = self.pe(inputs)
+        causal_mask = self.get_causal_attention_mask(inputs)
```
Comment (fchollet) [inline examples/vision/image_captioning.py:318]: I think you should take the OR of the padding mask (`mask` argument) and the causal mask, no? Otherwise you will pay attention to padded locations.

File: examples/vision/image_captioning.py [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+
+        attention_output_1 = self.attention_1(
+            query=inputs, value=inputs, key=inputs, attention_mask=causal_mask
+        )
+        out_1 = self.layernorm_1(inputs + attention_output_1)
+
+        attention_output_2 = self.attention_2(
+            query=out_1, value=encoder_outputs, key=encoder_outputs, attention_mask=None
```
Comment (fchollet) [inline examples/vision/image_captioning.py:326]: You need to provide the padding mask here, no?

Chances are this will improve your results. Sorry for missing this earlier.

File: examples/vision/image_captioning.py [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+        self.attention = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = layers.Dense(embed_dim, activation="relu")
+        self.layernorm_1 = layers.LayerNormalization()
+
+    def call(self, inputs, training, mask=None):
+        inputs = self.dense_proj(inputs)
+        attention_output = self.attention(
+            query=inputs, value=inputs, key=inputs, attention_mask=None
```
Comment (fchollet) [inline examples/vision/image_captioning.py:261]: We should also pass the padding mask here

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Things are looking great. I pushed some copyedits.

I believe we should leverage the padding mask in the Transformer model (sorry for not catching this earlier)

**Comment by AakashKumarNain** [comment]
Thanks for the thorough review @fchollet I will make the necessary changes

File: examples/vision/image_captioning.py (hunk 17 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:261]: Is it really required here though? The model should be look at all the image features at once when applying attention to gather the global context for caption, right?

File: examples/vision/image_captioning.py (hunk 16 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:326]: No worries Francois. I will try adding it

File: examples/vision/image_captioning.py (hunk 15 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:318]: Ok. Let me update that

File: examples/vision/image_captioning.py [hunk 18]
```diff
@@ -0,0 +1,510 @@
...
+
+
+def read_image(img_path, size=IMAGE_SIZE):
+    img = tf.io.read_file(img_path)
+    img = tf.image.decode_jpeg(img, channels=3)
+    img = tf.image.resize(img, IMAGE_SIZE)
+    return img
+
+
+def text_to_seq(cap):
```
Comment (fchollet) [inline examples/vision/image_captioning.py:201]: Also you don't really need this utility function since you can just do `.map(vectorization)` (saves you 2 lines of code)

File: examples/vision/image_captioning.py (hunk 17 above)
Comment (fchollet) [reply examples/vision/image_captioning.py:261]: That's a good point, there's no padding going on with the input sequence here so no masking necessary.

Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:261]: Thanks for confirming

File: examples/vision/image_captioning.py (hunk 18 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:201]: Ah yes!

File: examples/vision/image_captioning.py (hunk 15 above)
Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:318]: What about this one? Because the encoder output is not padded at all, do we still need to use the `padding_mask`?

Comment (fchollet) [reply examples/vision/image_captioning.py:318]: The padding mask (`mask` argument) here refers to the `inputs` sequence which will need padding.

You'd also need to pass the `mask` argument here:

```
batch_seq_pred = self.decoder(batch_seq_inp, encoder_out, training=training)
```

And take the mask into account when computing the loss and accuracy as well.

Comment (fchollet) [reply examples/vision/image_captioning.py:318]: Both the causal mask and the padding mask refer to the same sequence (the caption), that's why they need to be combined

Comment (AakashKumarNain) [reply examples/vision/image_captioning.py:318]: Thanks @fchollet for the detailed feedback. Yeah, I see now that it is a mistake in the code. Apologies for that. Here is what I have figured out:

```python

# Pass this mask to the decoder block
mask = tf.cast(tf.math.not_equal(batch_seq_inp,  0), tf.int32)  # (batch_size, seq_len)

# Compute the causal mask in the attention block on PE output
causal_mask = get_causal_attention_mask(positional_embed_out) # (batch_size, seq_len, seq_len)

# Compute the padding mask
padding_mask = tf.cast(mask[:, tf.newaxis, :], tf.int32) # (batch_size, 1, seq_len)

# Take the minimum of two
padding_mask = tf.minimum(causal_mask, padding_mak) # (batch_size, seq_len, seq_len)

# This mask has to be passed to the attention layer where encoder output and the first attention layer
# output is combined
attention_output_2 = self.attention_2(
            query=out_1,
            value=encoder_outputs,
            key=encoder_outputs,
            attention_mask=padding_mask,
        )
```

But there is a problem
```
Shape of query (out_1) -> (B, 39, 512)
Shape of value (encoder_out) -> (B, 100, 512)
Shape of padding mask - > (B, 39, 39)
```

For the attention layer to work as expected, the padding mask should have a shape of `(B, 39, 100)`. Should we pad the padding_mask to make sure it has the correct shape or do you have another solution to propose?

File: examples/vision/image_captioning.py [hunk 19]
```diff
@@ -0,0 +1,510 @@
...
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.dense_proj = keras.Sequential(
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim)]
+        )
+        self.layernorm_1 = layers.LayerNormalization()
+        self.layernorm_2 = layers.LayerNormalization()
+        self.layernorm_3 = layers.LayerNormalization()
+
+        self.pe = PositionalEmbedding(
```
Comment (fchollet) [inline examples/vision/image_captioning.py:310]: Call this `self.embedding`, variable names should always stay very readable

File: examples/vision/image_captioning.py (hunk 15 above)
Comment (fchollet) [reply examples/vision/image_captioning.py:318]: The minimum mask (combined) has to be passed to `attention_1`.

As for `attention_2`... for a given sample, `mask[i, j]` indicates the interaction between `caption[i]` and `image[j]`. We want that to be 0 for `i > caption_length` so you need to pass the padding mask. I think shape `(B, 39, 1)` should work. No need to pass the combined mask at that step, I think, because that's only relevant when the caption interacts with itself, no when the caption interacts with the image.

**Comment by fchollet** [comment]
So did it visibly improve your results?

**Comment by AakashKumarNain** [comment]
> So did it visibly improve your results?

I am trying to converge it. It overfits very quickly at this point, and dropout in the encoder/decoder doesn't seem to be helping much. Interestingly, it's not typical overfitting, but more like a mode collapse. So, I will double check the code again, will run a fe more experiments and will report back. 

Thanks @fchollet for all the support.

**Comment by AakashKumarNain** [comment]
@fchollet I made some changes in the code and it's finally working as expected. Let me know what you think

**Review by fchollet — COMMENTED** [review]
Excellent! Congrats on getting everything running! 👍 

I pushed some last copyedits. Please pull them and generate the ipynb and md files. Let's ship this!

**Comment by AakashKumarNain** [comment]
Thank you @fchollet for all the help. I have added all the necessary files.

PS: This was a fun exercise!

**Review by fchollet — APPROVED** [review]
Thanks a lot for the contribution! This is an awesome addition to keras.io.
